sch_cake.c

```diff
@@ -1619,12 +1619,20 @@ static unsigned int cake_drop(struct Qdisc *sch, struct sk_buff **to_free)
 	return idx + (tin << 16);
 }
 
+static inline __be16 cake_skb_proto(const struct sk_buff *skb)
+{
+	if (skb->protocol == htons(ETH_P_8021Q))
```

Comment:
Not sure if this is enough for offloaded VLANs? See vlan_proto() in br_netfilter_hooks, which does this:
```
	if (skb_vlan_tag_present(skb))
		return skb->protocol;
	else if (skb->protocol == htons(ETH_P_8021Q))
		return vlan_eth_hdr(skb)->h_vlan_encapsulated_proto;
```

Reply:
Yes, I take br_netfilter_hooks function as source of inspiration, but in cake's case it doesn't really matter whether VLAN inserted into skb's data or exists as vlan_tci in skb head - in both cases cake_handle_diffserv() needs to know real ethernet protocol, no matter of vlan offloading type.

br_netfilter_hooks' vlan_proto() will return 0 if no vlan is used also, which is obviously wrong for sch_cake, so first and the last branches in vlan_proto() in our case must return skb->protocol, just as I did in patch.

Reply:
Hmm, yeah, guess you're right that it'll do the right thing in both cases...